import "ydb/library/actors/protos/actors.proto";
import "ydb/core/protos/base.proto";

package NKikimrStateStorage;
option java_package = "ru.yandex.kikimr.proto";

// todo: add followers info

message TEvLookup {
    optional fixed64 TabletID = 1;
    optional uint64 Cookie = 2;
};

message TEvInfo {
    optional NKikimrProto.EReplyStatus Status = 1;
    optional fixed64 TabletID = 2;
    optional uint64 Cookie = 3;
    optional NActorsProto.TActorId CurrentLeader = 4;
    optional NActorsProto.TActorId CurrentLeaderTablet = 5;
    optional uint32 CurrentGeneration = 6;
    optional uint32 CurrentStep = 7;
    optional bool Locked = 8;
    optional uint64 LockedFor = 9;
    optional uint64 Signature = 10;
    repeated NActorsProto.TActorId Follower = 11;
    repeated NActorsProto.TActorId FollowerTablet = 12;
    repeated NActorsProto.TActorId FollowerCandidates = 13;
    optional fixed32 ConfigContentHash = 14;
};

message TEvReplicaShutdown {
};

message TEvDumpRequest {
};

message TEvDump {
    repeated TEvInfo Info = 1;
};

message TEvUpdate {
    optional fixed64 TabletID = 1;
    optional uint64 Cookie = 2;
    optional NActorsProto.TActorId ProposedLeader = 3;
    optional NActorsProto.TActorId ProposedLeaderTablet = 4;
    optional uint32 ProposedGeneration = 5;
    optional uint32 ProposedStep = 6;
    optional uint64 Signature = 7;
    optional bool IsGuardian = 8;
};

message TEvDelete {
    optional fixed64 TabletID = 1;
};

message TEvCleanup {
    optional fixed64 TabletID = 1;
    optional NActorsProto.TActorId ProposedLeader = 2;
}

message TEvRegisterFollower {
    optional fixed64 TabletID = 1;
    optional NActorsProto.TActorId Follower = 2;
    optional NActorsProto.TActorId FollowerTablet = 3;
    optional bool Candidate = 4;
}

message TEvUnregisterFollower {
    optional fixed64 TabletID = 1;
    optional NActorsProto.TActorId Follower = 2;
}

message TEvLock {
    optional fixed64 TabletID = 1;
    optional uint64 Cookie = 2;
    optional NActorsProto.TActorId ProposedLeader = 3;
    optional uint32 ProposedGeneration = 4;
    optional uint64 Signature = 5;
};

message TEvReplicaLeaderDemoted {
    optional fixed64 TabletID = 1;
    optional uint64 Signature = 2;
};

message TEvReplicaBoardPublish {
    optional string Path = 1;
    optional bytes Payload = 2;
    optional uint64 TtlMs = 3;
    optional bool Register = 4;
    optional NActorsProto.TActorId Owner = 5;
};

message TEvReplicaBoardPublishAck {
};

message TEvReplicaBoardLookup {
    optional string Path = 1;
    reserved 2;
    optional bool Subscribe = 3;
};

message TEvReplicaBoardCleanup {
};

message TEvReplicaBoardUnsubscribe {
};

message TBoardEntryInfo {
    optional NActorsProto.TActorId Owner = 1;
    optional bytes Payload = 2;
    optional bool Dropped = 3 [default = false];
};

message TEvReplicaBoardInfo {
    optional string Path = 1;
    optional bool Dropped = 2;
    repeated TBoardEntryInfo Info = 3;
};

message TEvReplicaBoardInfoUpdate {
    optional string Path = 1;
    optional TBoardEntryInfo Info = 3;
};

message TEndpointBoardEntry {
    optional string Address = 1;
    optional uint32 Port = 2;
    optional float Load = 3;

    optional bool Ssl = 4;
    repeated string Services = 5;

    optional string DataCenter = 6;
    optional uint32 NodeId = 7;

    repeated string AddressesV4 = 8;
    repeated string AddressesV6 = 9;
    optional string TargetNameOverride = 10;
};

